### DISCUSSION TITLE
unable to resolve external zone url from a pod running inside the cluster

### DISCUSSION THREAD

**Discussion by micksey9** [coredns/coredns#4879] [Q&A] [answered]
Hey all, I'm fairly new to coredns and I need some clarification regarding k8s_external plugin. Following is my corefile config:

  _Corefile: |
    .:53 {
        log
        errors
        health {
            lameduck 5s
        }
        ready
        kubernetes k8cluster.local in-addr.arpa ip6.arpa {
          pods insecure
          fallthrough in-addr.arpa ip6.arpa
        }
        k8s_external example.org {
           ttl 30
        }
        prometheus :9153
        forward . /etc/resolv.conf {
          prefer_udp
        }
        cache 30
        loop
        reload
        loadbalance
    }_

I have k8s_external plugin configured to serve external zone **example.org**. My understanding is that if I create a service with an external ip I should be able to resolve it not only from outside the cluster but also from any pod running inside the cluster. e.g the following service:

**NAME         TYPE           CLUSTER-IP     EXTERNAL-IP    PORT(S)          AGE
**my-service**   LoadBalancer   10.233.44.98   **10.41.12.139**   8080:31712/TCP   29m**

I'm able to resolve from the physical master node: 
**[root@node1 test]# dig @10.233.0.3 my-service.default.example.org +short
10.41.12.139**

but when I try to do an nslookup for the same from inside a pod it fails with the following error:

**[root@node1 test]# kubectl exec -it grafana -n monitoring -- nslookup my-service.default.example.org**
Defaulted container "grafana" out of: grafana, init-chown-data (init)
Server:         169.254.25.10
Address:        169.254.25.10:53

**** server can't find my-service.default.example.org: NXDOMAIN

** server can't find my-service.default.example.org: NXDOMAIN**

Am I wrong in my expectation for the the resolution to work for **my-service.default.example.org** from inside a pod ??? 
Also not sure if its relevant but in the resolv.conf on the pod I only see my cluster local domains and the external domains that are part of my /etc/resolv.conf file on the physical node.

**Comment by chrisohaver** [comment, marked as answer]
It looks like you have node local dns deployed. node local dns puts a coredns instance on each node that forwards cluster zone queries to the cluster coredns, and everything else directly to upstream dns. 

In this case, the queries for example.com are forwarded directly upstream, and never reaching the cluster coredns. 

The best solution I think would be to configure the node local dns instances to forward example.com to the cluster coredns service (adding to the node local coredns config, a server block for example.com that forwards to cluster coredns cluster ip).

**Reply by micksey9** [reply]
Thanks a lot @chrisohaver, changing nodelocaldns config resolved the issue. 
Is it possible to point this out in k8s_external plugin documentation perhaps?? That if we have nodelocaldns configured then we need to make config changes there as well to get the plugin to work and be able to resolve external url from the pod?
